src/neighbors.cc

```diff
@@ -775,25 +777,23 @@ peerDigestLookup(CachePeer * p, HttpRequest * request)
 }
 
 /* select best CachePeer based on cache digests */
-CachePeer *
-neighborsDigestSelect(HttpRequest * request)
+void
+neighborsDigestSelect(PeerSelector *selector, HttpRequest * request)
 {
-    CachePeer *best_p = NULL;
 #if USE_CACHE_DIGESTS
 
-    int best_rtt = 0;
-    int choice_count = 0;
-    int ichoice_count = 0;
-    CachePeer *p;
     int p_rtt;
     int i;
+    int choice_count = 0;
 
     if (!request->flags.hierarchical)
-        return NULL;
+        return;
 
     storeKeyPublicByRequest(request);
 
-    for (i = 0, p = first_ping; i++ < Config.npeers; p = p->next) {
+    std::vector<std::pair<int, CachePeer *> > sortedPeers;
+    i = 0;
+    for (CachePeer *p = first_ping; i++ < Config.npeers; p = p->next) {
```

Comment:
The first_ping is required here. Actually does a round-robin on peers list. 

It is also used when pinging peers (getNeighborsToPing function) where If we remove it then always (mostly) the first peer will be used because will be pinged first so it will respond first.
I can see this behaviour in my tests, using virtual machines on a remote server. 

Inside neighborsDigestSelect used to start peerDigestLookup from different peer in each PeerSelection procedure.  This is works well. The key in this loop is the following code:
````
  if (i == 1)
            first_ping = p;
````
However maybe it is better to not use the first_ping (which is used to getNeighborsToPing function) but use a new first_digest_loopup variable to give us a better round-robin.

Reply:
> The key in this loop is the following code:

AFAICT, that code does nothing:

```
i = 0;
for (CachePeer *p = first_ping; i++ < Config.npeers; p = p->next) {
if (i == 1)
            first_ping = p;
```

... is equivalent to this code:

```
i = 0;
for (CachePeer *p = first_ping; i++ < Config.npeers; p = p->next) {
```

because, during the first loop iteration and only during the first loop iteration, `i` is `1` and `p` is `first_ping`.

You can ignore this concern because the code changed so much. I will look at the new/revamped code with your getNeighborsToPing comments in mind.

Reply:
True. This is fixed on my latest commit